src/hotspot/share/utilities/bitMap.inline.hpp

```diff
@@ -188,46 +188,140 @@ inline BitMap::idx_t BitMap::get_next_bit_impl(idx_t l_index, idx_t r_index) con
+    // Flipped and shifted first word is non-zero.
+    idx_t result = l_index + count_trailing_zeros(cword);
+    if (aligned_right || (result < r_index)) {
+      return result;
+    }
+
+    // Result is beyond range bound
+    return r_index;
+  }
+
+  // Flipped and shifted first word is zero.  Word search through
+  // aligned up r_index for a non-zero flipped word.
+  idx_t word_limit = aligned_right
+      ? to_words_align_down(r_index) // Miniscule savings when aligned.
+      : to_words_align_up(r_index);
+
+  // Check the rest
+  while (++word_index < word_limit) {
+    cword = word(word_index, flip);
+    if (cword != 0) {
+      idx_t result = bit_index(word_index) + count_trailing_zeros(cword);
+      if (aligned_right || (result < r_index)) {
+        return result;
       }
-      // No bits in range; return r_index.
+      // Result is beyond range bound; return r_index.
+      assert((word_index + 1) == word_limit, "invariant");
+      return r_index;
     }
   }
+
+  // No bits in range
   return r_index;
 }
 
+static BitMap::idx_t high_order_bit_index(BitMap::bm_word_t cword) {
+  return ((BitsPerWord - 1) - count_leading_zeros(cword));
+}
+
+template<BitMap::bm_word_t flip, bool aligned_left>
+inline BitMap::idx_t BitMap::get_prev_bit_impl(idx_t l_index, idx_t r_index_exclusive) const {
+  STATIC_ASSERT(flip == find_ones_flip || flip == find_zeros_flip);
+  verify_range(l_index, r_index_exclusive);
+  assert(!aligned_left || is_aligned(l_index, BitsPerWord), "l_index not aligned");
+
+  if (l_index == r_index_exclusive) {
+    // Empty range
+    return idx_t(-1);
+  }
+
+  // The first word often contains an interesting bit, either due to
+  // density or because of features of the calling algorithm.  So it's
+  // important to examine that first word with a minimum of fuss,
+  // minimizing setup time for later words that will be wasted if the
+  // first word is indeed interesting.
+
+  // Get the word containing r_index, and shift out high bits.
+  idx_t r_index = r_index_exclusive - 1;
+  idx_t word_index = to_words_align_down(r_index);
+  idx_t r_index_in_word = bit_in_word(r_index);
+  idx_t r_index_bit = size_t(1) << r_index_in_word;
+
+  bm_word_t cword_unmasked = word(word_index, flip);
+
+  // Check first bit
+  if ((cword_unmasked & r_index_bit) != 0) {
+    // The first bit is similarly often interesting. When it matters
+    // (density or features of the calling algorithm make it likely
+    // the first bit is set), going straight to the next clause compares
+    // poorly with doing this check first; count_leading_zeros can be
+    // relatively expensive, plus there is the additional range check.
+    // But when the first bit isn't set, the cost of having tested for
+    // it is relatively small compared to the rest of the search.
+    return r_index;
+  }
+
+  // Mask out bits not part of the search; excluding first bit
+  idx_t cword_mask = r_index_bit - 1;
+  idx_t cword = cword_unmasked & cword_mask;
```

Comment:
Types are wrong here - both should be `bm_word_t`.  `cword` has the correct type in the forward search.